File: 966-binary-subarrays-with-sum/binary-subarrays-with-sum.py
class Solution:
    def numSubarraysWithSum(self, nums: List[int], goal: int) -> int:
        def helper(nums,goal):
            l,r,sum1=0,0,0
            count=0
            n=len(nums)
            while(r<n):
                sum1+=nums[r]
                while(sum1>goal and l<=r):
                    sum1-=nums[l]
                    l+=1
                if(sum1<=goal):
                    count+=r-l+1
                r+=1    
            return count
        return helper(nums,goal)-helper(nums,goal-1)                

        

File: 966-binary-subarrays-with-sum/binary-subarrays-with-sum.java
class Solution {
    public int numSubarraysWithSum(int[] nums, int goal) {
        return helper(nums,goal)-helper(nums,goal-1);
    }
    public  int helper(int[] nums,int goal){
            int l=0,r=0,count=0,sum=0;
            int n=nums.length;
            while(r<n){
                 sum+=nums[r];
                 while(sum>goal && l<=r){
                    sum-=nums[l];
                    l+=1;
                 }
                 if(sum<=goal){
                    count+=(r-l+1);
                 }
                 r+=1;              
            }
            return count;
        }
}